### PR TITLE
perf(provider): adaptive persistent-history Qwen3.6 MTP

### PR DESCRIPTION
**Pull request by Gajesh2007**
## Summary

Umbrella provider integration for the Qwen3.6 adaptive persistent-history MTP stack.

Depends on:

- https://github.com/Layr-Labs/mlx-swift-lm/pull/110 — merged as `244b104cb69cf90eb0cf79eb607a91dca0ac0af6`

This PR:

- enables request-stateful Qwen assistants to use adaptive rectangular depth `k=0...4`
- keeps stateless Gemma on its existing fixed behavior
- propagates EngineV2's resolved recurrent residency into the provider's process-wide admission budget
- reserves 4F for reachable compact `k >= 2` transitions, while captured `k <= 1` remains at the base 3F
- exposes adaptive state through the existing MTP status/metrics surface
- extends the real Qwen3.6 production-bundle canary across text, vision, tools, concurrency, deterministic parity, cancellation, and capacity drain
- updates the provider changelog

No coordinator or wire-protocol changes. The `mtp` beta remains default-off.

## Imported optimization stack

The merged dependency supplies:

1. **Hierarchical Metal top-2** — two exact GPU reductions over target logits; no full vocabulary sort or intermediate host synchronization.
2. **Persistent proposal-head KV** — complete committed history remains request-owned, and leading history appends K/V without computing dead output rows.
3. **Adaptive depth** — one rectangular `k=0...4` per scheduler plan from request-local conditional acceptance and shared measured marginal cost.
4. **Compact recurrent replay** — widths `S >= 3` retain transformed GatedDelta inputs and the final state; full acceptance lazily detaches the exact conv tail, strict-prefix acceptance replays only that prefix, and rejection rolls back without tail-copy kernels.
5. **Exact admission** — provider and engine consume the same resolved recurrent and assistant residency figures, including strict/full transition roots retained until the successor commit fence.

For conditional acceptance probabilities `p_i`, the expected emitted-token count is:

`E[emitted(k)] = 1 + Σ(i=1...k) Π(j=1...i) p_j`

Extra draft depth is selected only while its expected accepted-token mass pays for incremental proposal and verification cost.

## Review hardening

The final review pass closes four reachable edge cases:

- fixed-depth stateful MTP no longer depends on the optional adaptive top-2 policy capability
- scheduler token-budget fallback is a replacement seed, preserving persistent proposal-head carry and recording the true fallback provenance
- strict-prefix and lazy full-accept backing roots remain charged until successor commit
- compact `k >= 2` provider reservations propagate the corrected 4F engine charge instead of the under-reserved 3F value

## Performance

Real Qwen3.6-35B-A3B MLX production artifact on an M4 Max. Each run decodes the same deterministic 192-token sequence, with three target-only and three MTP repetitions; token IDs must match exactly.

| Run | Target-only median | MTP median | Speedup | Wall-time reduction |
|---|---:|---:|---:|---:|
| 1 | 7.657386 s | 3.813989 s | 2.0077x | 50.2% |
| 2 | 7.457212 s | 3.791629 s | 1.9668x | 49.2% |

Observed result: **1.97x-2.01x**, or approximately **2x decode throughput with 49-50% lower decode wall time**.

Important attribution: this is active MTP versus target-only decoding in the same final build. It validates the complete resulting stack, but it is not a commit-parent versus PR microbenchmark and is DEBUG canary evidence, not release-throughput evidence.

## Behavior and safety

- generated tokens remain target-authoritative
- greedy MTP output is byte/token identical to target-only output in the live canary
- full acceptance commits the exact recurrent final state and retains lazy-copy backing until a safe fence
- strict-prefix acceptance replays exactly the accepted transformed-input prefix
- rejection/cancellation discards speculative state and releases all reservations
- only committed target transitions enter persistent proposal-head history
- widths 1/2 retain captured-state verification and the base 3F charge
- multimodal Qwen rows remain target-only where media-position hidden capture is unavailable
- fixed/serial fallbacks remain supported and correctly charged

## Before

```mermaid
flowchart LR
  subgraph Behavior
    A1[Qwen request] --> B1[Fixed k=1 proposal]
    B1 --> C1[Target verify width 2]
    C1 --> D1[Commit captured state or rollback]
  end
  subgraph Code
    E1[EngineV2SlotFactory] --> F1[fixedDraftTokens = 1]
    F1 --> G1[CBv2 captured recurrent stack]
    H1[Provider admission] --> I1[Re-derived base recurrent bytes]
  end
```

## After

```mermaid
flowchart LR
  subgraph Behavior
    A2[Qwen request] --> H2[Request-owned proposal-head history]
    H2 --> P2[Adaptive k=0...4 proposal]
    P2 --> V2[One rectangular target verification]
    V2 --> O2{Outcome}
    O2 -->|Full| F2[Lazy exact-tail detach + backing charge]
    O2 -->|Strict prefix| R2[Compact replay + tape charge]
    O2 -->|Reject| X2[Rollback without replay or tail copy]
    F2 --> N2[Successor commit releases transition roots]
    R2 --> N2
  end
  subgraph Code
    S2[EngineV2SlotFactory] --> Q2[Request-stateful capability]
    Q2 --> C2[CBv2MTPDepthController]
    C2 --> E2[RecurrentStateV2 transition retention]
    E2 --> B2[EngineV2 resolved 3F or 4F charge]
    B2 --> G2[ProductionBuild and EngineV2Bridge admission]
  end
```

## Verification

Parent commit `328383efa` pins merged dependency commit `244b104cb69cf90eb0cf79eb607a91dca0ac0af6`.

The merged dependency commit has the same Git tree as the fully reviewed and tested dependency head `72ce217796e39ef44d0a2a509f9d27c0e8659e99`; `git diff --stat` and `git diff --check` between them are empty.

Provider repository:

- full provider suite: **82 XCTest cases**, 0 failures
- full Swift Testing suite: **2,092 tests across 213 suites**, 0 failures
- focused exact fixed-byte propagation test — passed
- real-artifact Qwen production-bundle canary — passed twice
- pre-push checks — passed
- `git diff --check` — passed

Dependency repository:

- `swift build -c release` — passed
- full package suite: **517 XCTest cases**, 1 skipped, 0 failures
- full Swift Testing suite: **858 tests across 115 suites**, 0 failures
- Qwen request-stateful MTP suite: **28 tests**, 0 failures
- compact recurrent capture/replay suite: **12 tests**, 0 failures
- independent performance and maintainability re-reviews — approved with no remaining actionable issue

New regressions specifically prove fixed-mode activation without adaptive top-2, production planner fallback/carry reuse, strict-prefix continuation within 4F, full-accept backing retention through successor commit, and downstream 4F propagation.

A release-mode **test** invocation remains blocked by the unrelated existing `SSDPrefixCacheTests.swift` reference to the DEBUG-only `holdDestructiveEpochForTesting` helper. The production release build itself succeeds.

### CODE REVIEW

**Comment by Gajesh2007** [comment]
CI status note: the code-bearing checks completed so far are green, including **Provider Tests** (9m11s), Coordinator Tests/Lint, Console UI, Prompt Sidecar, and Release Integrity. `Threat Model Review` failed because fork-origin PRs receive an empty `ANTHROPIC_API_KEY`; the three Vercel checks report organization authorization required. Those are fork-secret/authorization failures, not test failures. E2E Integration and E2E Benchmarks are still pending.

**Comment by Gajesh2007** [comment]
### 1,024-token follow-up

Ran the real Qwen3.6-35B-A3B production-bundle canary with an exact **1,024 generated-token** requirement. The test asserted that target-only reached all 1,024 tokens, ran three target-only and three MTP repetitions, and required every MTP token ID to match target-only.

| Generation | Target-only median | MTP median | Speedup | Wall-time reduction |
|---:|---:|---:|---:|---:|
| 1,024 tokens | 37.328035 s | 20.284353 s | **1.8402x** | **45.7%** |

That is **84.0% higher decode throughput**. The complete text/vision/tools/concurrency/parity/cancellation/capacity canary passed in 184.988 seconds.

Same attribution caveat: MTP versus target-only in the same final DEBUG build on M4 Max; this is not an isolated parent-versus-PR or release-throughput benchmark.
